scene/2d/gpu_particles_2d.cpp

```diff
@@ -28,12 +28,92 @@
 /* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
 /*************************************************************************/
 
+//connect("changed",
+
 #include "gpu_particles_2d.h"
 
 #include "scene/resources/particles_material.h"
 
 #ifdef TOOLS_ENABLED
 #include "core/config/engine.h"
+
+void GPUParticles2D::set_material(const Ref<Material> &p_material) {
+	if (material_in_use != nullptr) {
+		material_in_use->disconnect("changed", callable_mp(this, &GPUParticles2D::_material_changed));
+		vis_rect_hvframes = Vector2(1, 1);
+	}
+	if (get_material().is_valid()) {
+		get_material()->disconnect("changed", callable_mp(this, &GPUParticles2D::_material_changed));
+		vis_rect_hvframes = Vector2(1, 1);
+	}
+
+	material_in_use = nullptr;
+	CanvasItem::set_material(p_material);
+
+	if (get_material().is_valid()) {
+		get_material()->connect("changed", callable_mp(this, &GPUParticles2D::_material_changed));
+	}
+	update();
+}
+
+void GPUParticles2D::_material_changed() {
+	if (material_in_use != nullptr) {
+		_calc_vis_rect_hvframes(material_in_use);
+	} else if (get_material().is_valid()) {
+		CanvasItemMaterial *mat = Object::cast_to<CanvasItemMaterial>(get_material().ptr());
+		_calc_vis_rect_hvframes(mat);
+	} else {
+		vis_rect_hvframes = Vector2(1, 1);
+	}
+	update();
+}
+
+void GPUParticles2D::_calc_vis_rect_hvframes(CanvasItemMaterial *p_material) {
+	if (p_material && p_material->get_particles_animation()) {
+		vis_rect_hvframes = Vector2(p_material->get_particles_anim_h_frames(), p_material->get_particles_anim_v_frames());
+	} else {
+		vis_rect_hvframes = Vector2(1, 1);
+	}
+}
+
+void GPUParticles2D::set_use_parent_material(bool p_use_parent_material) {
+	if (p_use_parent_material) {
+		if (material_in_use != nullptr) {
+			material_in_use->disconnect("changed", callable_mp(this, &GPUParticles2D::_material_changed));
+			vis_rect_hvframes = Vector2(1, 1);
+		}
+		if (get_material().is_valid()) {
+			get_material()->disconnect("changed", callable_mp(this, &GPUParticles2D::_material_changed));
+			vis_rect_hvframes = Vector2(1, 1);
+		}
+
+		CanvasItem::set_use_parent_material(p_use_parent_material);
+		CanvasItem *ci = Object::cast_to<CanvasItem>(this);
+		while (ci && ci->get_use_parent_material()) {
+			Node *n = ci->get_parent();
+			ci = Object::cast_to<CanvasItem>(n);
+		}
+
+		material_in_use = Object::cast_to<CanvasItemMaterial>(ci->get_material().ptr());
```

Comment:
Couldn't this be a `Ref<CanvasItemMaterial>` instead of a raw pointer?

Reply:
Yes, changed! :)